cryptoki/src/session/object_management.rs

```diff
@@ -13,7 +13,217 @@ use std::convert::TryInto;
+                    )
+                },
+                None => {
+                    // C_FindObjects() is not implemented on this implementation
+                    // sort of unexpected. TODO: Consider panic!() instead?
+                    log::error!("C_FindObjects() is not implemented on this library");
+                    return Some(Err(Error::NullFunctionPointer) as Result<ObjectHandle>);
+                }
+            };
+
+            if let Rv::Error(error) = Rv::from(p11rv) {
+                return Some(
+                    Err(Error::Pkcs11(error, Function::FindObjects)) as Result<ObjectHandle>
+                );
+            }
+        }
+        None
+    }
+}
+
+impl Drop for ObjectHandleIterator<'_> {
+    fn drop(&mut self) {
+        // silently pass if C_FindObjectsFinal() is not implemented on this implementation
+        // this is unexpected. TODO: Consider panic!() instead?
+        if let Some(f) = get_pkcs11_func!(self.session.client(), C_FindObjectsFinal) {
+            // swallow the return value, as we can't do anything about it
+            let _ = unsafe { f(self.session.handle()) };
+        }
+    }
+}
+
 impl Session {
+    /// Iterate over session objects matching a template.
+    ///
+    /// # Arguments
+    /// * `template` - The template to match objects against
+    ///
+    /// # Returns
+    ///
+    /// This function will return a [`Result<ObjectHandleIterator>`] that can be used to iterate over the objects
+    /// matching the template. Note that the cache size is managed internally and set to a default value (10)
+    ///
+    /// # See also
+    /// * [`ObjectHandleIterator`] for more information on how to use the iterator
+    /// * [`Session::iter_objects_with_cache_size`] for a way to specify the cache size
+    #[inline(always)]
+    pub fn iter_objects(&self, template: &[Attribute]) -> Result<ObjectHandleIterator> {
+        self.iter_objects_with_cache_size(template, MAX_OBJECT_COUNT)
+    }
+
+    /// Iterate over session objects matching a template, with cache size
+    ///
+    /// # Arguments
+    /// * `template` - The template to match objects against
+    /// * `cache_size` - The number of objects to cache. Note that 0 is an invalid value and will return an error.
+    ///
+    /// # Returns
+    ///
+    /// This function will return a [`Result<ObjectHandleIterator>`] that can be used to iterate over the objects
+    /// matching the template. The cache size corresponds to the size of the array provided to `C_FindObjects()`.
+    ///
+    /// # See also
+    /// * [`ObjectHandleIterator`] for more information on how to use the iterator
+    /// * [`Session::iter_objects`] for a simpler way to iterate over objects
+    pub fn iter_objects_with_cache_size(
+        &self,
+        template: &[Attribute],
+        cache_size: usize,
+    ) -> Result<ObjectHandleIterator> {
+        let template: Vec<CK_ATTRIBUTE> = template.iter().map(|attr| attr.into()).collect();
```

Comment:
```suggestion
        let template: Vec<CK_ATTRIBUTE> = template.iter().map(Into::into).collect();
```

I think this would be shorter :thinking: